packages/classes/src/PayloadArray.js

```diff
@@ -0,0 +1,98 @@
+import Payload from './Payload'
+
+/**
+ * PayloadArray stores a series of Payloads in a data structure that allows for easy filtering
+ * and lookup by position and type.
+ */
+class PayloadArray {
+	payloads = []
+
+	constructor(...payloads) {
+		[...payloads].forEach((payload) => this.insert(payload))
+	}
+
+	get length() { return this.payloads.length }
+
+	/**
+	 * Insert a Payload into the PayloadArray while maintaining position order.
+	 *
+	 * @param  {Payload} payload The Payload to be inserted.
+	 * @return {PayloadArray}    This PayloadArray to allow for chaining.
+	 */
+	insert = (payload) => {
+		if (!(payload instanceof Payload)) {
+			throw new Error('Attempt to insert a non-Payload to PayloadArray')
+		}
+		const index = this.indexOfPosition(payload.position)
+		this.payloads.splice(index, 0, payload)
+		return this
+	}
+
+	/**
+	 * Find the index of the first Payload whose position is greater than or equal to this one.
+	 *
+	 * NOTE: if no qualifying payloads exist, this will return the length of the PayloadArray
+	 *
+	 * @param  {Number} position The position being searched for.
+	 * @return {Number}          The index of the item, or the PayloadArray length.
+	 */
+	indexOfPosition = (position) => {
```

Comment:
I don't guess it matters too much, but I find this method to be fairly illegible. I think it would benefit from a single comment above each `return` explaining what we've determined at this point?

Reply:
Recursion is like time travel, it's not meant to be truly understood.

(I'll see what I can do)

Reply:
OK so in seriousness: the most legible / simple solution (linear time) would be to just go through the array and stop when we get an item that is higher than the passed position.

This would be much easier to parse, but would be less efficient than this recursive binary-search-inspired approach.  I feel like optimization is probably quite important since this is a core data type (and it's likely that payloads will be fairly high volume), but I also appreciate the value of legibility.

Reply:
I wonder if this is one of those times it'd make more sense to use a dep like [lodash.sortedIndex](https://www.npmjs.com/package/lodash.sortedindex) (note: broken out, does not come with whole lodash library) instead of embedding our own implementation of a search in here.

Reply:
I assumed it all was _purposeful_, that's why my suggestion was to improve commenting rather than rewrite.

I trust your judgment on _using_ this method, I just literally can't read it (my eyes keep crossing), and thus can't really inspect it for possible bugs/shortcomings. Granted, I don't know what binary searches are, and probably most other devs do.

If this is basically a JS implementation of a common method, I can just go read about that method. But it still might be good to decorate the comment with something like "this takes a binary-search-inspired approach for optimization" so that folks like me can go "oh okay, if I really want to know what it's doing I'll go learn about binary searches first".

(To be clear, I'm _not_ asking that our yard stick for all legibility be an English major who stumbled into programming. I'm just saying that's the only perspective I have!)

Reply:
Absolutely!  And I'm not pushing back on the feedback, just thinking through some options and pointing out that the tradeoff itself can be questioned too.

@chriszs I admit when writing it I thought of the adage of "if you're implementing quicksort you're doing something wrong".

Reply:
I don't always recommend one function packages, but when I do they implement binary search.

Reply:
I think Chris's suggestion is the right one... and that this has literally already been implemented and we should just import that one thing.  I think this will be the first time I've ever supported importing something from lodash!

just to be clear lodash is the not evil one (or was that underscore).  Phewf.

Reply:
Lodash is the evil one.

Reply:
Looking into this a bit more -- one concern is that [lodash.sortedindexby](https://www.npmjs.com/package/lodash.sortedindexby) (which is the flavor I think we would want) is still pretty heavy compared to this implementation.  2k+ lines of code vs the 15 or whatever we have here.  Going to sleep on it a bit but I'm starting to think that just making a utility implementation of sortedIndexBy would be a reasonable solution here.  Use the name, use the same API, just implemented with a little less rigor...  honestly I'm having trouble understanding what could possibly require that much stuff.

Reply:
Update: it seems to be rooted in the number of "types" of thing that needs to be supported (e.g. comparing strings, paths, etc) -- since we know it's just a number I think that's what lets us simplify dramatically.

Reply:
What? Why? How? That's insane! sortedIndex is 209 lines, 117 of which are comments. How did adding an accessor method argument add 1,791 lines? On reviewing the code this was generated from, I can't tell. It doesn't look like it should add that many. But clearly I'm missing something.

Reply:
I have gone into the belly of the beast. There is no making heads or tails of how this happened. Flee! Save yourselves!

Reply:
[probably not going to use recursion](https://softwareengineering.stackexchange.com/questions/179863/performance-recursion-vs-iteration-in-javascript).

Reply:
Resolving this -- I re-implemented a stripped down version of the lodash sortdIndexBy, put it in a utility file, and I think that is our best balance for now.